src/calls/embedTemplatePage.ts

```diff
@@ -121,3 +111,98 @@ function createEmbedIframe(options: CreateIframeOptions): HTMLIFrameElement {
 
   return iframe;
 }
+
+function parseTokenPayload(token: string): SdkTokenPayload {
+  const tokenPayloadResult = sdkTokenPayloadSchema.safeParse(
+    JSON.parse(atob(token.split(".")[1] ?? "")),
+  );
+
+  if (!tokenPayloadResult.success) {
+    throw new Error("Invalid token payload");
+  }
+
+  return tokenPayloadResult.data;
+}
+
+function validatePermissions(
+  permissions: SdkPermission[],
+  features?: TemplateEmbedFeatures,
+): void {
+  const hasPermission = (validPermissionSets: SdkPermission[][]) => {
+    return validPermissionSets.every((permissionSet) =>
+      permissionSet.some((permission) => permissions.includes(permission)),
+    );
+  };
+
+  if (!hasPermission(REQUIRED_PERMISSIONS.base)) {
+    throw new Error(
+      "workspace:read, user:read, and template:read permissions are required",
+    );
+  }
+
+  if (!features) {
+    return;
+  }
+
+  if (features.passList && !hasPermission(REQUIRED_PERMISSIONS.passList)) {
+    throw new Error("passList feature requires pass:read permission");
+  }
+
+  if (
+    features.templateEditor &&
+    !hasPermission(REQUIRED_PERMISSIONS.templateEditor)
+  ) {
+    throw new Error(
+      "templateEditor feature requires template:write permission",
+    );
+  }
+
+  if (features.campaigns && !hasPermission(REQUIRED_PERMISSIONS.campaigns)) {
+    throw new Error("campaigns feature requires campaign:read permission");
+  }
+
+  if (
+    features.campaignEditor &&
+    !hasPermission(REQUIRED_PERMISSIONS.campaignEditor)
+  ) {
+    throw new Error(
+      "campaignEditor feature requires campaign:write permission",
+    );
+  }
```

Comment:
I think we could take this one step further and loop over the keys of `features` here. Not technically safe but probably okay:
```suggestion
  const failedPermissions = Object.entries(features)
    .filter(([_, enabled]) => enabled)
    .map(([feature]) => ({feature, REQUIRED_PERMISSIONS[feature]}))
    .filter(({permissions}) => permissions !== undefined)
    .filter(({permissions}) => !hasPermission(permissions));

  if (failedPermissions.length !== 0) {
    throw new Error(
      failedPermissions
        .map(({feature, permission) => `${feature} feature requires ${permissions.join(" or ")} permission`)
        .join(", ")
    )
  }
```

Reply:
Tweaked the error messaging generation a little, but seems to work pretty well